recipes-ni/ni-base-system-image-tests/files/fs_permissions_known.py

```diff
@@ -0,0 +1,162 @@
+import argparse
+import os
+import sys
+import hashlib
+import stat
+import pwd
+import grp
+import fs_permissions_shared
+
+# Function to return tree instead of variable since it contains functions inside its structure
+def known_permissions_tree():
+    return {
+        '.': system_dir,
+        'boot': {
+            '.': system_dir,
+            'bootmode': system_file,
+            'grub': {
+                '.': system_dir,
+                '*': system_file,
+                'grubenv': permissions(0o664, 'admin', 'ni', FT_REG),
+                'grubenv.bak': permissions(0o664, 'admin', 'ni', FT_REG),
+                'recoverytool-ni-version': permissions(0o0444, 'admin', 'administrators', FT_REG)
+            }, 'runmode': {
+                '.': system_dir,
+                '*': system_file,
+                'bzImage': system_link(f'bzImage-{bash("$(uname -r)")[0]}'),
+            }
+        }, 'lib': {
+            '.': system_dir,
+            'modules': {
+                '.': system_dir,
+                '$(uname -r)': {
+                    '.': system_dir,
+                    '**': system_hier,
+                    'source': system_link('build')
+                }
+            }
+        }
+    }
+
+# Use bash to resolve strings like path globs and $() expressions.
+def bash(path):
+    cmd = [ 'bash', '-O', 'extglob', '-c', f'echo {path}' ]
+    return fs_permissions_shared.run_cmd(cmd).strip().split(' ')
+
+# Convert the tree structure into a list of paths and functions to check their permissions
+def known_permissions():
+    def permissions_list(tree, path, perms):
+        for file in tree:
+            if type(tree[file]) is dict:
+                perms = permissions_list(tree[file], f'{path}/{file}', perms)
+            elif file == '.':
+                perms[bash(path if path != '' else '/')[0]] = tree[file]
+            else:
+                subfiles = bash(f'{path}/{file}')
+                for subfile in subfiles:
+                    perms[subfile] = tree[file]
+        return perms
+    return permissions_list(known_permissions_tree(), '', {})
+
+#------- Functions for the tree to check file permissions ------
+
+FT_DIR = 0
+FT_REG = 1
+FT_LNK = 2
+
+# Report errors when a kind of thing is different than expected
+def log_mismatch(kind, exp, act, path, logger, md5sum):
+    md5sum.update(f'{kind}: {exp} != {act}'.encode('utf-8'))
+    logger.log(f'ERROR: Expected {kind} \'{exp}\' but got \'{act}\' for path {path}.')
+
+# Main check for file/directory permissions
+def permissions(mode, user, group, file_type):
+    def ret(path, stats, logger, md5sum):
+        nonlocal mode
+        nonlocal user
+        nonlocal group
+        nonlocal file_type
```

Comment:
formatting nitpick: `nonlocal` [can use a comma-separated list of identifiers](https://docs.python.org/3/reference/simple_stmts.html?highlight=nonlocal#the-nonlocal-statement)

```suggestion
        nonlocal mode, user, group, file_type
```